bin/runtime-common/src/messages.rs

```diff
@@ -1183,4 +1194,56 @@ mod tests {
 			.collect()),
 		);
 	}
+
+	#[test]
+	fn verify_messages_proof_with_parser_does_not_panic_if_messages_count_mismatches() {
+		assert_eq!(
+			target::verify_messages_proof_with_parser::<OnThisChainBridge, _, _>(
+				(
+					Default::default(),
+					StorageProof::new(vec![]),
+					Default::default(),
+					0,
+					u64::MAX
+				),
+				0,
+				|_, _| Ok(TestMessageProofParser {
+					failing: false,
+					messages: 0..=u64::MAX,
+					outbound_lane_data: Some(OutboundLaneData {
+						oldest_unpruned_nonce: 1,
+						latest_received_nonce: 1,
+						latest_generated_nonce: 1,
+					}),
+				}),
+			),
+			Err(target::MessageProofError::MessagesCountMismatch),
+		);
+	}
+
+	#[test]
+	#[should_panic]
+	fn verify_messages_proof_with_parser_panic_if_too_many_messages_declared() {
```

Comment:
I don't think so. The problem before was that `messages_in_the_proof` would have been calculated to `0`. This allowed to bypass the triggering of the `MessagesCountMismatch` error in case `message_count` was `0` as well. But then for the actual allocation of the vec the saturated value of `end - begin`, i.e. `u64::MAX` would have been used. 

Now you could still bypass the `MessageCountMismatch` error but in this case an allocation of size `0` would be done. This prevents the unbound allocation.

Reply:
> IDK if it is possible to just return Ok(ProvenMessages::new()) right way in case the checked_sub() overflows instead of continuing with messages_in_the_proof == 0 which in the end should result in the same thing, i.e. no proven messages.

That would be wrong, because empty proof is not allowed (`MessageProofError::Empty`). So we can't simply return `Ok(ProvenMessages::new())` - instead we just skip "messages part" and go to the "outbound lane data" part of the function.